### PR TITLE
Support WSEGSICD rebase

### PR DESCRIPTION
**Pull request by GitPaean**
This is the rebased version of https://github.com/OPM/opm-common/pull/344 . 

Tests have been added. And more adjustments might be needed due to the new thoughts or the update of the implementation of other parts. 

A lot of previous discussion should be at OPM/opm-parser#1194. .

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2178 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2178 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2178 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2178 please

File: opm/parser/eclipse/EclipseState/Schedule/MSW/Segment.hpp [hunk 1]
```diff
@@ -100,6 +112,13 @@ namespace Opm {
         // indicate if the data related to 'INC' or 'ABS' is ready
         // the volume will be updated at a final step.
         bool m_data_ready;
+        // indicate the type of the segment
+        // regular or spiral ICD
+        SegmentType m_segment_type = SegmentType::REGULAR;
+
+        // information related to SpiralICD. It is nullptr for segments are not
+        // spiral ICD type
+        std::shared_ptr<SpiralICD> m_spiral_icd;
```
Comment (GitPaean) [inline opm/parser/eclipse/EclipseState/Schedule/MSW/Segment.hpp]: WSEGSICD will be developed along with WSEGVALV for now .  Due to the time limit, we might not be able to get a base class ICD for them.  When we have a third ICD type join, definitely we should have a ICD base class, then here will be 
```
std::shared_prt<ICD> m_icd;
```

The tricky part will be the interface and what if ICD types changes. 

Everything is still in progress, let us see what will happen.

**Comment by GitPaean** [comment]
This PR is ready for review.  I will address the comment about enum for status.

**Comment by GitPaean** [comment]
jenkins build this om-simulators=2178 please

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 2]
```diff
@@ -319,6 +319,9 @@ namespace {
         else if (keyword.name() == "COMPSEGS")
             handleCOMPSEGS(keyword, currentStep, grid, parseContext, errors);
 
+            else if (keyword.name() == "WSEGSICD")
```
Comment (atgeirr) [inline src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: Indentation.

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 3]
```diff
@@ -411,6 +414,7 @@ namespace {
             } else {
                 std::string msg = "OPM does not support grid property modifier " + keyword.name() + " in the Schedule section. Error at report: " + std::to_string( currentStep );
                 parseContext.handleError( ParseContext::UNSUPPORTED_SCHEDULE_GEO_MODIFIER , msg, errors );
+
```
Comment (atgeirr) [inline src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: Added extra blank?

File: src/opm/parser/eclipse/EclipseState/Schedule/MSW/WellSegments.cpp [hunk 4]
```diff
@@ -417,6 +417,45 @@ namespace Opm {
...
+        } else {
+            // other segments
+            const int outlet_segment_number = segment.outletSegment();
+            const Segment &outlet_segment = getFromSegmentNumber(outlet_segment_number);
+
+            segment_length = segment.totalLength() - outlet_segment.totalLength();
+        }
+
+        if (segment_length <= 0.)
+            throw std::runtime_error(" non positive segemnt length is obtained for segment "
```
Comment (atgeirr) [inline src/opm/parser/eclipse/EclipseState/Schedule/MSW/WellSegments.cpp]: Typo: segemnt -> segment

File: tests/parser/WellTests.cpp [hunk 5]
```diff
@@ -216,6 +216,7 @@ BOOST_AUTO_TEST_CASE(isProducerCorrectlySet) {
     }
 
 
+
```
Comment (atgeirr) [inline tests/parser/WellTests.cpp]: Extra blanks added to this file for no reason.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by bska** [comment]
In the downstream/restart portion of this (PR #1272), @joakim-hove [queried](https://github.com/OPM/opm-common/pull/1272#discussion_r354334353) about the unit conversion support for the SICD strength parameter, and wondered whether we should add a new unit dimension to simplify that calculation.  As a preparation for doing so I had a look at the datamodel for the [`WSEGSICD`](https://github.com/OPM/opm-common/blob/1534570b3d509087071f9b7e9676108188906c05/src/opm/parser/eclipse/share/keywords/000_Eclipse100/W/WSEGSICD) keyword, and in particular at the [specification](https://github.com/OPM/opm-common/blob/1534570b3d509087071f9b7e9676108188906c05/src/opm/parser/eclipse/share/keywords/000_Eclipse100/W/WSEGSICD#L5) for the unit of measurement for this parameter.

I can't quite understand that unit.  It looks to be based on geometric volume rates, but it really should be
```
            [Pressure]
-------------------------------------
 [Reservoir Rate] * [Reservoir Rate]
```
e.g.,
```
psia / (rb/day)**2
```
in the case of FIELD units.  Using `Length` here happens to do the right thing for METRIC, LAB, and PVT-M, but it will generate a suspicious value for FIELD.

Am I missing something?

**Comment by joakim-hove** [comment]
> Am I missing something?

Probably not; the whole unit system  - and in particular plethora of dimensions often makes me slightly uncomfortable.

**Comment by bska** [comment]
> > Am I missing something?
> 
> Probably not; the whole unit system - and in particular plethora of dimensions often makes me slightly uncomfortable.

Actually, I'm wrong here.  It really should be based on geometric volume rates so `[Length]**3/[Time]` is the correct rate unit for this quantity.  I'll still add a new dimension to represent it though so I don't run into that error again.  Sorry about the noise.

**Comment by GitPaean** [comment]
@bska the discussion was here, 

https://github.com/OPM/opm-parser/pull/1181

And some other bug is fixed here, 

https://github.com/OPM/opm-common/pull/1203

**Pull request merged**